### PR TITLE
Fix string comparisons.

### PR DESCRIPTION
**Pull request by amrishlal**
## Description
A query such as `SELECT count(*) FROM mytable WHERE trim(OriginState) = DestState` currently throws exception, because internally the predicate gets rewritten to `MINUS(trim(OriginState), DestState) = 0` and MINUS function fails to evaluate STRING arguments. This PR modifies `StringPredicateFilterOptizer.java` to properly rewrite predicate to use STRCMP instead of MINUS (`STRCMP(trim(OriginState), DestState) = 0`).

Note that the same issue was fixed for string columns in #7394. This PR is extending the original fix to include string functions as well.
## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/statement/StringPredicateFilterOptimizer.java [hunk 1]
```diff
@@ -99,21 +104,37 @@ private static void replaceMinusWithCompareForStrings(Expression expression, Sch
...
+  /** List of string functions that return STRING output. */
+  private static Set<String> getStringOutputFunctionList() {
+    Set<String> set = new HashSet<>();
+    Method[] methods = StringFunctions.class.getDeclaredMethods();
+    for (Method method : methods) {
+      if (method.getReturnType() == String.class) {
+        set.add(method.getName().toUpperCase());
+      }
+    }
+    return set;
```
Comment (richardstartin) [inline pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/statement/StringPredicateFilterOptimizer.java:138]: `@ScalarFunction` now has a `names` array parameter - so you need to read the names from the annotation to do this correctly.

Comment (amrishlal) [reply pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/statement/StringPredicateFilterOptimizer.java:138]: Wouldn't using @ScalarFunction return names of all scalar functions and how would the output type of the function be resolved? We just want String scalar functions and that too only the once that return a string as output. All the string scalar functions are declared in StringFunctions and hence this code.

**Comment by amrishlal** [comment]
> Please remove the binary file added by accident.

Done

File: pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/statement/StringPredicateFilterOptimizer.java (hunk 1 above)
Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/statement/StringPredicateFilterOptimizer.java:138]: Sorry, allow me to clarify:

```java
    Set<String> set = new HashSet<>();
    Method[] methods = StringFunctions.class.getDeclaredMethods();
    for (Method method : methods) {
      if (method.getReturnType() == String.class) {
        if (method.isAnnotationPresent(ScalarFunction.class)) {
           ScalarFunction annotation = method.getAnnotation(ScalarFunction.class);
           for (String name : annotation.names()) {
             set.add(name.toUpperCase());
           }   
        }
        set.add(method.getName().toUpperCase());
      }
    }
    return set;
```

This way your logic works even if there is an alias in use. Does that make sense?

Comment (amrishlal) [reply pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/statement/StringPredicateFilterOptimizer.java:138]: Hmm, I see. Thanks for clarifying.

**Review by richardstartin — APPROVED** [review]
👍🏻

**Pull request merged**